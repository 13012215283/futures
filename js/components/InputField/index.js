import React, { Component } from 'react';
import { StyleSheet, TextInput } from 'react-native';
import InputFieldStyle from './inputField.style';

// @flow

const InputFieldStyles = StyleSheet.create(InputFieldStyle);
const noop = () => {};
type Props = {
  type: string,
  inputStyle?: Object,
  onChange: Function,
  onBlur: Function,
  keyboardType: string,
};

type State = {
  text: string,
};
export default class InputField extends Component<Props, State> {
  static defaultProps = {
    type: 'text',
    inputStyle: {},
    onChange: noop,
    onBlur: noop,
    keyboardType: 'numeric',
  };

  onChange(text) {
    const { onChange } = this.props;

    if (onChange) {
      onChange(text);
    }
  }

  render() {
    const { inputStyle } = this.props;
    return (
      <TextInput
        {...this.props}
        underlineColorAndroid="transparent"
        onChange={e => {
          this.onChange(e.nativeEvent.text);
        }}
        style={[InputFieldStyles.input, inputStyle]}
        placeholder="请输入"
      />
    );
  }
}
