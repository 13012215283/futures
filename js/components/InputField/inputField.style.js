import { StyleSheet } from 'react-native';
import colors from '../themes/colors';
import { px2dp } from '../../utils/px2dp';

export default {
  input: {
    width: px2dp(250),
    height: px2dp(60),
    borderWidth: StyleSheet.hairlineWidth,
    padding: 0,
    borderColor: colors['1103'],
    fontSize: px2dp(28),
    color: colors['1101'],
    textAlign: 'center',
  },
};
