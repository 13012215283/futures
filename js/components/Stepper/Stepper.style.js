import { px2dp } from '../../utils/px2dp';
import { colors } from '../themes';

export default {
  stepWrap: {
    flexDirection: 'row',
  },
  stepButton: {
    width: px2dp(60),
    height: px2dp(60),
    borderWidth: 1,
    borderColor: colors['1001'],
    justifyContent: 'center',
    backgroundColor: '#rgba(232,59,59,0.2)',
  },
  stepText: {
    textAlign: 'center',
    color: colors['1001'],
    fontSize: px2dp(32),
  },
  inputItem: {
    marginLeft: px2dp(14),
    marginRight: px2dp(14),
  },
  disabledButton: {
    backgroundColor: 'rgba(189,189,189,0.2)',
    borderColor: colors['1103'],
    borderWidth: 1,
  },
  disabledText: {
    color: colors['1102'],
  },
  warningInput: {
    borderColor: '#ff0000',
  },
};
