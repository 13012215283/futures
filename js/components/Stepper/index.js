import React, { Component } from 'react';
import { View, TouchableWithoutFeedback, Text, StyleSheet } from 'react-native';
import equal from 'fast-deep-equal';

import { formatVal } from '../../utils/numberFormat';
import InputField from '../InputField';
import style from './Stepper.style';

const styles = StyleSheet.create(style);
const MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || 2 ** 53 - 1;

// @flow
type Props = {
  step?: number | string,
  min?: number,
  max?: number,
  value?: string,
  originalValue?: string,
  inputViewStyle?: Object,
  inputItemStyle?: Object,
  onChange?: Function,
  onEndEditing?: Function,
  editable?: boolean,
};

export default class Stepper extends Component<Props> {
  static defaultProps = {
    step: 0.01,
    max: MAX_SAFE_INTEGER,
    min: -MAX_SAFE_INTEGER,
    inputViewStyle: {},
    editable: true,
  };

  /* 数字精度 */
  static getPrecision(num) {
    const numStr = num.toString();
    let precision = 0;
    if (numStr.indexOf('.') >= 0) {
      precision = numStr.length - numStr.indexOf('.') - 1;
    }
    return precision;
  }

  constructor(props) {
    super(props);
    this.state = {
      value: props.value,
      max: props.max,
      min: props.min,
      isWarning: false,
      editable: !!props.editable,
    };
    this.onChange = this.onChange.bind(this);
    this.stepDown = this.stepDown.bind(this);
    this.stepUp = this.stepUp.bind(this);
    this.onEndEditing = this.onEndEditing.bind(this);
    this.setValue = this.setValue.bind(this);
  }

  // componentDidMount() {
  //   Keyboard.addListener('keyboardDidHide', () => {
  //     const { value } = this.state;
  //     const { max, min } = this.props;
  //     this.setState({
  //       isWarning: !(value <= max && value >= min),
  //     });
  //   });
  // }

  componentWillReceiveProps(nextProps) {
    this.setState({
      max: nextProps.max,
      min: nextProps.min,
      editable: nextProps.editable || true,
    });
  }

  shouldComponentUpdate(nextProps, nextState) {
    return !equal(nextState, this.state);
  }

  onChange(value) {
    const toValue = formatVal(value, ',');
    const { onChange, max } = this.props;
    this.checkInput(toValue);
    this.setState({
      value: `${Number(toValue) > Number(max) ? max : toValue}`,
    });

    if (onChange) {
      onChange(toValue);
    }
  }

  onEndEditing(e) {
    const { text } = e.nativeEvent;
    const { max, min, onEndEditing } = this.props;
    this.setState({
      isWarning: !(text < max && text > min),
    });

    if (onEndEditing) {
      onEndEditing(text);
    }
  }

  getValue() {
    return this.state.value;
  }

  setValue(value, callback = () => {}) {
    this.setState(
      {
        value,
      },
      callback
    );
  }

  checkInput(value) {
    const { min } = this.props;
    this.setState({
      isWarning: !(value >= min),
    });
  }

  computeValue(action: 'up' | 'down') {
    const direct = action === 'up' ? 1 : -1;
    const { step, originalValue } = this.props;
    const { value } = this.state;
    const precision = Stepper.getPrecision(step);

    const parseVal = Number(value);
    const parseStep = Number(step);
    const parseOriginalPrice = Number(originalValue);
    const reminder = Math.abs(parseOriginalPrice - value) % parseStep;
    let result;
    const closedValue =
      (parseOriginalPrice - parseVal) * direct > 0
        ? direct * reminder
        : direct * (10 - reminder);

    if (originalValue && !Number.isNaN(parseOriginalPrice) && reminder !== 0) {
      result = (parseVal + closedValue).toFixed(precision);
    } else {
      result = (parseVal + direct * parseStep).toFixed(precision);
    }
    this.onChange(result);
  }

  stepDown() {
    if (this.state.value > this.props.min) {
      this.computeValue('down');
    }
  }

  stepUp() {
    if (this.state.value < this.props.max) {
      this.computeValue('up');
    }
  }

  render() {
    const value = +this.state.value;
    const { isWarning, max, min, editable } = this.state;
    const { inputViewStyle, inputItemStyle } = this.props;
    let upDisabledStyle = null;
    let downDisabledStyle = null;
    let upDisabledTextStyle = null;
    let downDisabledTextStyle = null;
    if (value >= max || !editable) {
      upDisabledStyle = style.disabledButton;
      upDisabledTextStyle = style.disabledText;
    }
    if (value <= min || !editable) {
      downDisabledStyle = style.disabledButton;
      downDisabledTextStyle = style.disabledText;
    }

    return (
      <View style={style.stepWrap}>
        <TouchableWithoutFeedback onPressIn={this.stepDown}>
          <View style={[styles.stepButton, downDisabledStyle]}>
            <Text style={[styles.stepText, downDisabledTextStyle]}>-</Text>
          </View>
        </TouchableWithoutFeedback>
        <View style={[styles.inputItem, inputViewStyle]}>
          <InputField
            type="money"
            value={this.state.value}
            onChange={this.onChange}
            thousandSeparator=","
            inputStyle={isWarning ? styles.warningInput : inputItemStyle}
            editable={editable}
            // onEndEditing={this.onEndEditing}
          />
        </View>
        <TouchableWithoutFeedback onPressIn={this.stepUp}>
          <View style={[styles.stepButton, upDisabledStyle]}>
            <Text style={[styles.stepText, upDisabledTextStyle]}>+</Text>
          </View>
        </TouchableWithoutFeedback>
      </View>
    );
  }
}
