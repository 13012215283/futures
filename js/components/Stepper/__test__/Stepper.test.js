import React from 'react';
import { TouchableWithoutFeedback } from 'react-native';
import TestRenderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Stepper from '../index';
import InputField from '../../InputField/index';

const stepper = shallow(
  <Stepper max={0.03} min={0.01} step={0.01} value="0.02" />
);

test('render correctly', () => {
  const tree = TestRenderer.create(<Stepper />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('render correctly', () => {
  const tree = TestRenderer.create(
    <Stepper max={10} min={5} step={1} value="1" />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('按钮数量', () => {
  const buttons = stepper.find(TouchableWithoutFeedback);
  expect(buttons.length).toBe(2);
});

test('数值增减及临界值', () => {
  const downStep = stepper.find(TouchableWithoutFeedback).at(0);
  const upStep = stepper.find(TouchableWithoutFeedback).at(1);

  downStep.simulate('pressIn');
  expect(stepper.state('value')).toBe('0.01');
  downStep.simulate('pressIn');
  expect(stepper.state('value')).toBe('0.01');
  upStep.simulate('pressIn');
  expect(stepper.state('value')).toBe('0.02');
  upStep.simulate('pressIn');
  expect(stepper.state('value')).toBe('0.03');
  upStep.simulate('pressIn');
  expect(stepper.state('value')).toBe('0.03');
});

test('输入框', () => {
  const input = stepper.find(InputField).at(0);
  input.simulate('change', '0.03');
  expect(stepper.state('value')).toBe('0.03');
  input.simulate('change', '0.04');
  expect(stepper.state('isWarning')).toBe(false);
  input.simulate('change', '0.00');
  expect(stepper.state('isWarning')).toBe(true);
});
