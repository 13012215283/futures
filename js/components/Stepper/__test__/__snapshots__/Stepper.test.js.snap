// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#rgba(232,59,59,0.2)",
          "borderColor": "#e83b3b",
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 60,
        },
        null,
      ]
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#e83b3b",
            "fontSize": 32,
            "textAlign": "center",
          },
          null,
        ]
      }
    >
      -
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "marginLeft": 14,
          "marginRight": 14,
        },
        Object {},
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={true}
      inputStyle={Object {}}
      keyboardType="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="请输入"
      style={
        Array [
          Object {
            "borderColor": "#bdbdbd",
            "borderWidth": 0.5,
            "color": "#212121",
            "fontSize": 28,
            "height": 60,
            "padding": 0,
            "textAlign": "center",
            "width": 250,
          },
          Object {},
        ]
      }
      thousandSeparator=","
      type="money"
      underlineColorAndroid="transparent"
      value={undefined}
    />
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#rgba(232,59,59,0.2)",
          "borderColor": "#e83b3b",
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 60,
        },
        null,
      ]
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#e83b3b",
            "fontSize": 32,
            "textAlign": "center",
          },
          null,
        ]
      }
    >
      +
    </Text>
  </View>
</View>
`;

exports[`render correctly 2`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#rgba(232,59,59,0.2)",
          "borderColor": "#e83b3b",
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 60,
        },
        Object {
          "backgroundColor": "rgba(189,189,189,0.2)",
          "borderColor": "#bdbdbd",
          "borderWidth": 1,
        },
      ]
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#e83b3b",
            "fontSize": 32,
            "textAlign": "center",
          },
          Object {
            "color": "#757575",
          },
        ]
      }
    >
      -
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "marginLeft": 14,
          "marginRight": 14,
        },
        Object {},
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={true}
      inputStyle={Object {}}
      keyboardType="numeric"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="请输入"
      style={
        Array [
          Object {
            "borderColor": "#bdbdbd",
            "borderWidth": 0.5,
            "color": "#212121",
            "fontSize": 28,
            "height": 60,
            "padding": 0,
            "textAlign": "center",
            "width": 250,
          },
          Object {},
        ]
      }
      thousandSeparator=","
      type="money"
      underlineColorAndroid="transparent"
      value="1"
    />
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#rgba(232,59,59,0.2)",
          "borderColor": "#e83b3b",
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 60,
        },
        null,
      ]
    }
    testID={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#e83b3b",
            "fontSize": 32,
            "textAlign": "center",
          },
          null,
        ]
      }
    >
      +
    </Text>
  </View>
</View>
`;
